
# Methylator package
**Methylator** is a bunch of `bash` & `R` scripts designed for DNA methylation studies with [Bisulfite sequencing](https://en.wikipedia.org/wiki/Bisulfite_sequencing) technics. It has been developped at the [BiBs](https://parisepigenetics.github.io/umr7216bioinfofacility/) plateform at [UMR7216 - Paris Epigenetics](http://parisepigenetics.com/) by Olivier Kirsh. An integration within a snakemake workflow is under development.          
  
**Methylator** performs the secondary analyses like genomic features annotations and figures.   
These R scripts are executed on HPC infrasctructures like [ipop-up](https://reyjul.gitlab.io/documentation-ipop-up/) or [ifb](https://ifb-elixirfr.gitlab.io/cluster/doc/) equiped with [slurm](https://slurm.schedmd.com/documentation.html) scheduler.  

The primary analysis (from `.fastq` files to `.bam` files) must first be done with **WGBS-workflow** <!-- add the like when the doc is ready...-->. 
<!-- one cool little short project, create RDta and methextract from a count table -->

## Methylator can do :  
- DMC : Differentially Methylated CpG analysis.  
- DMT : Differentially Methylated Tiles analysis.   
- eDMR : Differentially Methylated Region analysis.   
- BEDGRAPH : Generate `.bedgraph` files (% methylation, coverage, differential % methylation and q-value)  
- FIGURES : Annotations and figures.  

**Methylator** is inspired and relies on 3 main recongnized R packages for DNA mehtylation analysis : 
 - [methylKit](https://github.com/al2na/methylKit)  
 - [eDRM](https://github.com/ShengLi/edmr)  
 - [annotatr](https://bioconductor.org/packages/release/bioc/html/annotatr.html)  
 
**Methylator** is executed within a Singularity/Apptainer container to respect the [FAIR](https://www.nature.com/articles/s41592-020-0742-y) principles. The "methylator.simg" contains r-base=4.1.3 and all its packages & dependencies installed with conda in an ubuntu environment. Apps version & image creation recipes are described bellow. 

## Methylator scripts & files  
Your analysis is controlled by `Methylator_xxx.sh` you can edit to speficy which analysis you want to do.  A `Methylator_Config.txt` must be edited to pass specific parameters to your analysis. Additional files like, `Annotatr_mm10_Customs.RData` can also be provided for user specific annotations or `custom_plot.R` for user-specific plots.  
<!-- `Annotatr_mm10.RData` must be revisited and a `plot.R`must be created   -->
<!-- I wonder if some of the information in `Methylator_xxx.sh` should be given in `Methylator_Config.txt` -->

### input files 
Your WGBS data are give to **Methylator** as an  `RData` file generated by **WGBS-workflow**.   

### Files to edit  
- `Methylator_xxx.sh`   
- `Methylator_Config.txt`  
- `custom_plot.R`   
- `Annotatr_mm10_Custom_V2.R` 

### Scripts (do not edit!)
- `MKitDMC_Analysis.R`  
- `MKitDMT_Analysis.R`  
- `MKit_EDMR_Analysis.R`  
- `MKit_Bedgraph.R`  
- `MKit_BedgraphDiff.R`   
- `WGBS_Fig.R`  
- `plot.R` (this file doesn't exist yet)  
- `Annotatr_mm10.R` 
- `Annotatr_mm10.RData`  
- `Annotatr_mm10_Customs.RData`   

### Singularity image
- `methylator.simg'

<!-- figures and annotatr stuff here C:\Users\olivi\Nextcloud\Bioinfo_Analyses\WGBSDROM_script\script\figures_V2  -->
<!-- annotation stuff here too C:\Users\olivi\Nextcloud\Bioinfo_Scripts_Tools\Annotations\Annot_with_mm10 -->


### Methylator Singularity container
All Apps, R and R packages are containerized in a Singularity image. This allows a strict version control and reproducibility of the analyses.  Please read the Singularity documentation or my (digest)[https://github.com/kirsho/Singularity/blob/master/Intro2Singularity.md] to learn more about Singularity.  
The `Singularity_Methylator` image can be recreated with this 2 files stored (here)[https://github.com/kirsho/DASH/tree/main/IMAGES/methylator] :  
- `Singularity`      
- `edmr_install.R`  



## Methylator files description
### Methylator_xxx.sh
Check this [example](https://github.com/kirsho/DASH/blob/main/WGBS/DMT_analysis/Methylator_DMT250_1_7.sh)
This is an sbatch file for **Methylator** WGBS secondary analysis.he one to rules them all and **must** be edited by user.  
- Slurm options header
- Analysis version
- Type of analyses <!-- what about exploration analyses??   -->
- Experimental design <!-- mainly works for merge, develop deplicates design   -->
- Input RData file(s) 

All bellow the skull must not be edited by user.

<!--Things to do: rethink how projec directories are created and mannaged.  -->
<!--Things to do: directories are created by default and not all used. add more ifesle depending of user choices  -->

### Methylator_Config.txt
<!-- this is not a real .txt file but an R scritp.  -->
<!--Things to do: make real config file with options defined in Methylator_xxx.sh & Methylator_Config.txt -->
Must be edited by user. It passes analyse threshold, values and parameters through the different script of the workflow.
